### ISSUE TITLE
Two finger zoom

### ISSUE THREAD

**Issue by Goodingn** [osmandapp/Osmand#3319]
When I zoom (in or out) using two fingers, when I lift my fingers off the screen, the display moves, so much that the object I was zooming into may not now be visible. This happens every time, making two finger zoom unuseable.

Cheers

**Comment by hakuchi** [comment]
for me zooming with two finger works perfect. 
Tested with OsmAnd from PlayStore and latest nightly.

You have to stop moving your fingers before lifting from screen. If you lift on finger and the other is still moving the display will move.

**Comment by Goodingn** [comment]
"You have to stop moving your fingers before lifting from screen"

Nope- zooming should be intuative. Also, this app isn't pannig and zooming, i.e. the map doesn't move while zooming, by moving the two fingers across the screen while zooming.

For the record, every time I've tried zooming, I've had both fingers immobile and lifted them off the screen as close to simultaneous as possible, but it always does some sort of move. It's an instant move, the screen pops to a new position.

I get that

**Comment by sonora** [comment]
I assume this is the same as #1393, and we can consolidate(?)

Personally, I have been trying to reproduce this on several devices, and I so far could not, which makes it likely this somehow depends on device and/or Android version.

**Comment by hakuchi** [comment]
Zooming in with two fingers is one option that OsmAnd support.

A other option is to zoom wit one finger.

And latest option is to use + & - on the screen(my preferred method).

None of this zooming options making problems to me.
Tested on two different Android 6.0 phones

**Comment by Goodingn** [comment]
Mine's Samsung A3 2016, 'droid 5.1.1

**Comment by Goodingn** [comment]
Osmand is the only app with this problem and I probably have 10 different mapping type apps with pinch zoom- some just zoom, some pan rotate and zoom simultaneosly with two fingers.

**Comment by Goodingn** [comment]
I've had another go at this. If I two finger zoom but leave the fingers on the screen for a few moments, then release, the zoom level changes a little, in a pop of the screen, but the center is correct. If I two finger zoom, but immediately (yet still precisely) lift fingers off, the center moves.

**Comment by arty-name** [comment]
This happens for me as well. I have noticed that the simplest way to reproduce it is like that: 

1. Put thumb on the screen
1. Move thumb slightly to drag the map a bit
1. Put index finger on the screen
1. Pinch fingers to zoom out
1. Lift thumb first
1. Lift index finger second

Looks like OsmAnd readjusts the coordinates as if it is thumb which is lifted last.

**Comment by sonora** [comment]
It is not reproducable at all on all devices I have accessible right now. But we are taking quite a bit of heat about it, also in google Play feedback, so somehow or somewhere this must be a real issue. Strange that so many users cannot reproduce, while others have the issue. Could be device or settings specific.

**Comment by arty-name** [comment]
if there’s a good tool for screen capturing I could try recording that

**Comment by sonora** [comment]
I have an idea: Can you please go to Settings / General settings and check if the Kaiman filter check box is on or off, and if changing it makes a difference?

**Comment by arty-name** [comment]
Nope, that did not help. The checkbox was checked, after unchecking it I still can reproduce the issue. However I have noticed a new detail for it. The jump actually happend even before I lift the index finger. And what might be more important, I have noticed that I do not actually lift thumb straight up, I bend it with a jerking movement, which might be sending the map to the side. As if a kinetics gone wrong.

**Comment by arty-name** [comment]
And yes, the point I have put my thumb on initially ends up under my index finger.

**Comment by sonora** [comment]
I guess it was not the Kaiman setting, but I can now see it, too. It is particularly pronounced for big zoom changes, and there needs to be a noticeable time span between taking the first and then the second finger off the screen, or else the issue does not show up, right?

**Comment by sonora** [comment]
@vshcherb  Yes, I can confirm it now. It looks like the screen is interpreting some gestures during the "lift the fingers" time period. Looks like in our code we have to make sure we suppress processing new gestures (like taps) until both fingers have been fully taken off.

**Comment by sonora** [comment]
Think my analysis was faulty. Your statement is to the point: When it happens, the place on the map where your thumb initially was jumps to under where your index finger in the end. Weird thing is it only happens very rarely for me, most of the time it works perfectly ok ...

**Comment by arty-name** [comment]
For me the important part is to move the first finger after putting it on the map. I assume the app goes into "move" mode, then the second finger arrives and the app does "zoom", but after zoom ends, the "move" mode still functions.

**Comment by sonora** [comment]
Yes, that's probaly it. We need to terminate the panning before starting to respond to the pinch zoom.

**Comment by sonora** [comment]
I was experimenting with it here but that fix did not seem to work: https://github.com/osmandapp/Osmand/commit/418976ffb14870619ce9f6320beeda070c532331

**Comment by sonora** [comment]
Quite a tough issue!  But I thin I managed to fix this, via commit  https://github.com/osmandapp/Osmand/commit/d9a222384827d3b24ccfa08bbdba49334d59359c. Can you please test the latest nightly?

- If the fix works, we would need to port it also the osmand-core.
- Our multiTouch support still does not support the concurrent combination of 2-finger zoom+rotate+drag, which would require a major code review. But I am not sure the missing "drag-while zoom/rotate" is an issue in practice.

**Comment by sonora** [comment]
Complete fix implemented, OsmAnd should now follow conventional touch screen **drag, rotate, and zoom** behavior!

**Comment by sonora** [comment]
PS: On the occasion I significantly overhauled our entire map touch interface a bit, fixing some smaller old bugs and artifacts. Think it behaves very well now, and I have researched it thoroughly, but it is quite complex, so please verify my testing. (Things affected include pinch zooming, double tap slide zooming, double tap zooming, two finger tap zooming, and short/long taps in connection with these operations.)

**Comment by arty-name** [comment]
Great job, @sonora! I cannot reproduce the bug anymore and the overall map interaction feels much better! Looking forward to fix being released.

**Comment by vshcherb** [comment]
Really nice job! I didn't find any bug yet :)))

**Comment by tleb** [comment]
Hope I'm not bringing a too bad news here, but I still have this problem. It does it if I lift either the first or the second pressed finger first.

Installed OsmAnd~ 2.5.4 from F-Droid on a OnePlus 3 on Android version 7.1.1.

Is it just that I don't have the right version?

**Comment by arty-name** [comment]
@tleb, you are probably seeing just the minor snap to the integer zoom level from the fractional one. I see it too. I don’t really know why the integer zoom level is a must, maybe it isn’t.

**Comment by Goodingn** [comment]
Osmand is v2.6.3 in the Play Store. Try the new version :)

**Comment by tleb** [comment]
I didn't update apps from F-Droid in a while, updating fixed it. Sorry guys.

There is a minor snap as you say @arty-name but it really is nothing.

**Comment by sonora** [comment]
Yes, I had put quite some work in last December and January, hope that 2.6.x now works flawlessly with respect to this.
